### PR TITLE
separating onnx parsing from burn-import

### PR DESCRIPTION
**Pull request by skewballfox**
### Checklist
- [ ] confirm we want the name to be onnx-ir
- [x] deduplicate and remove unused dependencies
- [x] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#1812

### Changes

Separated out all the onnx specific code in burn-import to a new crate `onnx-ir` (which we can rename to whatever), and made some necessary changes to burn-import since `OnnxGraph` and other ir structs are now external. I also aliased a few types to just better distinguish between onnx and burn types.

some code had to be duplicated (`flatten_config`)or made public for the sake of using between both (`constant_conversion`)

### Testing

so far I've tested with onnx-test and burn-import. All test seem to be passing. examples seem okay

I think we should rework how onnx test are being used so that the models, and test themselves could be potentially reused. I think we can do that separately from this PR though.

Python side that would mean serializing graph inputs and outputs, burn side I'm guessing some macros for op types (such as `binary_op_test!`) though I'm open to suggestions if you guys want to handle it another way.

### CODE REVIEW

File: crates/onnx-ir/Cargo.toml
Comment (antimora) [inline crates/onnx-ir/Cargo.toml:16]: Dead code. Also please check if all dependencies still needed.

File: crates/onnx-ir/Cargo.toml
Comment (antimora) [inline crates/onnx-ir/Cargo.toml:27]: Dead

**Review by antimora — COMMENTED** [review]
Looks good. A minor comment to review/clean up the dependencies. 

But I think it's almost ready to be approve. 

Since this is a brand new crate, we also need to update `publish.yml` under `.github.yml`. 

Also, can you please double check `adding-a-new-operation-to-burn.md` is up to date? Not sure if we have any references to about the onnx portion. 

@nathanielsimard @laggui can you sign off on the name?

**Comment by laggui** [comment]
`onnx-ir` sounds good to me! Will try to take some time to review this.

File: crates/onnx-ir/Cargo.toml
Comment (skewballfox) [reply crates/onnx-ir/Cargo.toml]: done

**Comment by skewballfox** [comment]
> Also, can you please double check `adding-a-new-operation-to-burn.md` is up to date? Not sure if we have any references to about the onnx portion.
 
I rewrote it, but I'm having trouble figuring out the best way to word it. I'd definitely appreciate a proof reader on the changed section. 

Also I added a readme though it's currently pretty barebones still figuring out what information it should have for other implementors.

**Review by laggui — COMMENTED** [review]
Totally forgot about this PR, I'll hold off on merging an new ONNX ops if we're going to merge this soon otherwise you'll always have to fix the conflicts 😅 

Otherwise, the refactor LGTM.

**Review by antimora — APPROVED** [review]
Approving in advance. There are a couple minor text fixes but looks good overall.

File: contributor-book/src/guides/adding-a-new-operation-to-burn.md
Comment (skewballfox) [reply contributor-book/src/guides/adding-a-new-operation-to-burn.md]: Formatted and changed the punctuation, wording

**Comment by antimora** [comment]
Merged! Thanks for your work. I hope this crate will be useful to other project as well.

**Pull request merged**
